controller/src/main/java/io/pravega/controller/eventProcessor/impl/EventProcessorCell.java

```diff
@@ -285,12 +289,18 @@ final void awaitTerminated() {
 
     private EventProcessor<T> createEventProcessor(final EventProcessorConfig<T> eventProcessorConfig) {
         EventProcessor<T> eventProcessor = eventProcessorConfig.getSupplier().get();
-        eventProcessor.checkpointer = (Position position) ->
-        checkpointStore.setPosition(process, readerGroupName, readerId, position);
+        eventProcessor.checkpointer = (Position position) -> {
+            checkpoint.set(position);
```

Comment:
It looks like these two lines should go in the other order. After all you don't want a race where the checkpoint is set but not set on the store.

Reply:
Actually there is a race but not what you are thinking of. Let me explain:

Actually the store is used for failover recovery by another process that identifies the reader has failed. Store is external (zookeeper) and may not be accessible (say after a session expiration) 

The local "checkpoint" field is for reader to self declare itself offline with the latest checkpoint identifying the position till which the event has been processed before this reader self declares itself offline. 

So the local field is independent of whether we have been able to update the store or not. If the reader shutsdown gracefully, the store does not come into picture.

Now coming to the race. In case of zookeeper session expiration, the controller also loses cluster membership and is declared as failed. Other controller instances will try to recover its work by calling its reader offline at the position from the store.
At the same time this reader is declaring itself offline with its local position. 
But that is fine, because that does not break our at least once processing guarantees even if we recover from an older checkpoint (a la store). 

What you are suggesting is also fine and will guarantee that we only ever use the checkpoints that have been successfully updated in the store. but i was going for slightly more optimal position.